src/Agent.Worker/AgentPluginManager.cs

```diff
@@ -0,0 +1,332 @@
+        private readonly Dictionary<Guid, Dictionary<string, Definition>> _supportedTasks = new Dictionary<Guid, Dictionary<string, Definition>>();
+
+        private readonly List<AgentPluginInfo> _taskPlugins = new List<AgentPluginInfo>()
+        {
+            new AgentPluginInfo()
+            {
+                AgentPluginAssembly = "Agent.RepositoryPlugin",
+                AgentPluginEntryPoint = "Agent.RepositoryPlugin.CheckoutTask"
+            }
+        };
+
+        private readonly List<AgentPluginInfo> _commandPlugins = new List<AgentPluginInfo>()
+        {
+            new AgentPluginInfo()
+            {
+                AgentPluginAssembly = "Agent.DropPlugin",
+                AgentPluginEntryPoint = "Agent.DropPlugin.ArtifactUploadCommand"
+            }
+        };
+
+        public Dictionary<string, Dictionary<string, AgentCommandPluginInfo>> SupportedLoggingCommands => _supportedLoggingCommands;
+        public Dictionary<Guid, Dictionary<string, Definition>> SupportedTasks => _supportedTasks;
+
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+
+            // Load task plugins
+            foreach (var plugin in _taskPlugins)
+            {
+                IAgentTaskPlugin taskPlugin = null;
+                string typeName = $"{plugin.AgentPluginEntryPoint}, {plugin.AgentPluginAssembly}";
+                AssemblyLoadContext.Default.Resolving += ResolveAssembly;
+                try
+                {
+                    Trace.Info($"Load task plugin from '{typeName}'.");
+                    Type type = Type.GetType(typeName, throwOnError: true);
+                    taskPlugin = Activator.CreateInstance(type) as IAgentTaskPlugin;
+                }
+                finally
+                {
+                    AssemblyLoadContext.Default.Resolving -= ResolveAssembly;
+                }
+
+                ArgUtil.NotNull(taskPlugin, nameof(taskPlugin));
+                ArgUtil.NotNull(taskPlugin.Id, nameof(taskPlugin.Id));
+                ArgUtil.NotNullOrEmpty(taskPlugin.Version, nameof(taskPlugin.Version));
+                if (!_supportedTasks.ContainsKey(taskPlugin.Id))
+                {
+                    _supportedTasks[taskPlugin.Id] = new Dictionary<string, Definition>(StringComparer.OrdinalIgnoreCase);
+                }
+
+                Trace.Info($"Loaded task plugin id '{taskPlugin.Id}' ({taskPlugin.Version}).");
+                _supportedTasks[taskPlugin.Id][taskPlugin.Version] = new Definition() { Directory = HostContext.GetDirectory(WellKnownDirectory.Work) };
+                _supportedTasks[taskPlugin.Id][taskPlugin.Version].Data = new DefinitionData()
+                {
+                    Author = taskPlugin.Author,
+                    Description = taskPlugin.Description,
```

Comment:
what about instance name format? who calculates the instance name format? is it the agent or the server? is it based on the instance name format defined in the task instance or in the task definition? have you thought through how this will work for plugin steps?

Reply:
i assume server will do it since that how other task been handled as well.